### PR TITLE
feat: add language detector middleware and helpers

### LINKED ISSUE #3785: Language Detector Middleware for Hono
## Background
While building an i18n system for Hono, I noticed the lack of a standardized way to detect user language preferences across different sources (URL, cookies, headers, etc.). This is a crucial component for internationalization, and currently, each developer needs to implement their own solution.

## Current Situation
Currently in Hono:
- No built-in way to detect user language preferences
- No standardized approach for language detection
- Developers need to write custom code for each project

## Proposed Solution
A language detector middleware that:
- Detects user language from multiple sources (URL path, query params, cookies, Accept-Language header)
- Provides type-safe access to detected language via `c.get('language')`
- Follows i18next's language detection patterns (widely adopted in the JS ecosystem)
- Is highly configurable yet simple to use
- Can be used as a foundation for other i18n-related Hono middlewares

Basic usage example:
```typescript
import { languageDetector } from 'hono/language'

app.use('*', languageDetector({
  supportedLanguages: ['en', 'ar', 'es'],
  fallbackLanguage: 'en'
}))

app.get('/', (c) => {
  const lang = c.get('language') // type-safe access
  return c.text(`Current language: ${lang}`)
})

### PR DESCRIPTION
**Pull request by lord007tn**
Closes #3785

## Description
Add language detector middleware to support future i18n functionality in Hono applications.

## Features
- Detect language from multiple sources:
  - Query parameters
  - Cookies
  - Accept-Language header
  - URL path
- Type-safe language access via `c.get('language')`
- Configurable detection order
- Support for language caching
- Comprehensive test coverage

## Usage
```typescript
import { languageDetector } from 'hono/language'

app.use('*', languageDetector({
  supportedLanguages: ['en', 'ar', 'es'],
  fallbackLanguage: 'en'
}))

app.get('/', (c) => {
  const lang = c.get('language')
  return c.text(`Current language: ${lang}`)
})

### The author should do the following, if applicable

- [X] Add tests
- [X] Run tests
- [X] `bun run format:fix && bun run lint:fix` to format the code
- [X] Add [TSDoc](https://tsdoc.org/)/[JSDoc](https://jsdoc.app/about-getting-started) to document the code
…function

### CODE REVIEW

**Comment by lord007tn** [comment]
if this got accepted am willing to add docs in the official website to let people know how to use it

**Comment by lord007tn** [comment]
i forgot to mention the full use of the middleware options are:
```typescript
app.use(
  '*',
  languageDetector({
    // Detection strategies and their order
    order: ['path', 'querystring', 'cookie', 'header'],

    // Supported languages and fallback
    supportedLanguages: ['en', 'ar'],
    fallbackLanguage: 'en',

    // Query parameter settings
    lookupQuerystring: 'lang', // ?lang=en

    // Path settings
    lookupFromPathIndex: 0, // /en/about -> takes 'en'

    // Cookie settings
    lookupCookie: 'user-language',
    caches: false, // Set to false to disable caching

    // Header settings
    lookupFromHeaderKey: 'accept-language',

    // Language code handling
    ignoreCase: true,
    convertDetectedLanguage: (lang: string) => {
      // Optional: Convert language codes
      // e.g., 'en-US' -> 'en'
      return lang?.split('-')[0]?.toLowerCase() ?? '';
    },

    // Debugging
    debug: process.env.NODE_ENV !== 'production',
  }),
);
```

File: src/middleware/language/language.ts [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  debug?: boolean
+}
+
+export interface LanguageVariables {
+  language: string
+}
+
+const DEFAULT_OPTIONS: DetectorOptions = {
+  order: ['querystring', 'cookie', 'header'],
+  lookupQuerystring: 'lang',
```
Comment (askorupskyy) [inline src/middleware/language/language.ts:52]:
```suggestion
  lookupQueryString: 'lang',
```

**Review by askorupskyy — COMMENTED** [review]
I love it! This entirely solves the lang detection for me. Would also be cool to have a way to disable cookie/query detection altogether.

File: src/middleware/language/language.ts (hunk 1 above)
Comment (lord007tn) [reply src/middleware/language/language.ts:52]: > I love it! This entirely solves the lang detection for me. Would also be cool to have a way to disable cookie/query detection altogether, or could this be done with `order` param?

yep just provide the order: ['header'] and it will only detect the header

**Comment by yusukebe** [comment]
Thanks @lord007tn I'll review this later.

Hey @sor4chi What do you think of this feature?

File: src/middleware/language/language.ts [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    httpOnly: true,
+  },
+  debug: false,
+}
+/**
+ * Parse Accept-Language header values with quality scores
+ * @param header Accept-Language header string
+ * @returns Array of parsed languages with quality scores
+ */
+export function parseAcceptLanguage(header: string): Array<{ lang: string; q: number }> {
```
Comment (sor4chi) [inline src/middleware/language/language.ts:73]: I feel that this function overlaps in functionality with [parseAccept](https://github.com/lord007tn/hono/blob/bdf1a309c5739fc9060dc45327d4135bfeded377/src/helper/accepts/accepts.ts#L20-L43). As an experiment, when I tried the following:

```ts
export function parseAcceptLanguage(header: string): Array<{ lang: string; q: number }> {
  return parseAccept(header)
    .map(({ type, q }) => ({ lang: type, q }))
    .sort((a, b) => b.q - a.q)
}
```

I was able to pass the tests (though I may have overlooked edge cases). In this case, how about implementing a common parser in the utils and having both `helper/accept` and `middleware/language` reference it?

**Review by sor4chi — COMMENTED** [review]
Thank you for the wonderful PR, I liked it. I have one suggestion, what do you think?

Also, I feel a bit of discomfort with the naming `hono/language`. Personally, I would prefer something like `hono/language-detector`, but I’ll leave it to @yusukebe.

**Comment by yusukebe** [comment]
Hi @lord007tn 

Sorry for my super late response! Let's prepare to ship this middleware in the next minor version, `v4.7`.

@sor4chi 

I also have concerns regarding the name and whether it should be `hono/language-detector` or `hono/language`. I know `language` is too general, but if the function imported from `hono/language` is `languageDetector`, the user can imagine the feature, so I think `hono/language` is ok. This means, as @lord007tn wrote above:

```ts
import { languageDetector } from 'hono/language'
```

File: src/middleware/language/language.ts (hunk 2 above)
Comment (yusukebe) [reply src/middleware/language/language.ts:73]: I also think the below @sor4chi 's idea is good.

> how about implementing a common parser in the utils and having both `helper/accept` and `middleware/language` reference it?

@lord007tn What do you think of this?

File: src/middleware/language/language.ts (hunk 1 above)
Comment (yusukebe) [reply src/middleware/language/language.ts:52]: @lord007tn 

Can you apply [the suggestion by @askorupskyy](https://github.com/honojs/hono/pull/3787/files#r1904874595)?

**Comment by yusukebe** [comment]
@lord007tn 

Can you apply the following diff to unify the coding style (using `const` for the function) with other code?

```diff
diff --git a/src/middleware/language/language.ts b/src/middleware/language/language.ts
index c298d3e9..e21f5a07 100644
--- a/src/middleware/language/language.ts
+++ b/src/middleware/language/language.ts
@@ -70,7 +70,7 @@ const DEFAULT_OPTIONS: DetectorOptions = {
  * @param header Accept-Language header string
  * @returns Array of parsed languages with quality scores
  */
-export function parseAcceptLanguage(header: string): Array<{ lang: string; q: number }> {
+export const parseAcceptLanguage = (header: string): Array<{ lang: string; q: number }> => {
   try {
     return header
       .split(',')
@@ -97,10 +97,10 @@ export function parseAcceptLanguage(header: string): Array<{ lang: string; q: nu
  * @param options Detector options
  * @returns Normalized language code or undefined
  */
-export function normalizeLanguage(
+export const normalizeLanguage = (
   lang: string | null | undefined,
   options: DetectorOptions
-): string | undefined {
+): string | undefined => {
   if (!lang) {
     return undefined
   }
@@ -126,7 +126,7 @@ export function normalizeLanguage(
 /**
  * Detects language from query parameter
  */
-export function detectFromQuery(c: Context, options: DetectorOptions): string | undefined {
+export const detectFromQuery = (c: Context, options: DetectorOptions): string | undefined => {
   try {
     const query = c.req.query(options.lookupQuerystring)
     return normalizeLanguage(query, options)
@@ -138,7 +138,7 @@ export function detectFromQuery(c: Context, options: DetectorOptions): string |
 /**
  * Detects language from cookie
  */
-export function detectFromCookie(c: Context, options: DetectorOptions): string | undefined {
+export const detectFromCookie = (c: Context, options: DetectorOptions): string | undefined => {
   try {
     const cookie = getCookie(c, options.lookupCookie)
     return normalizeLanguage(cookie, options)
@@ -238,7 +238,7 @@ function cacheLanguage(c: Context, language: string, options: DetectorOptions):
 /**
  * Detect language from request
  */
-function detectLanguage(c: Context, options: DetectorOptions): string {
+const detectLanguage = (c: Context, options: DetectorOptions): string => {
   let detectedLang: string | undefined

   for (const detectorName of options.order) {
@@ -277,7 +277,7 @@ function detectLanguage(c: Context, options: DetectorOptions): string {
  * @param userOptions Configuration options for the language detector
  * @returns Hono middleware function
  */
-export function languageDetector(userOptions: Partial<DetectorOptions> = {}): MiddlewareHandler {
+export const languageDetector = (userOptions: Partial<DetectorOptions>): MiddlewareHandler => {
   const options: DetectorOptions = {
     ...DEFAULT_OPTIONS,
     ...userOptions,
@@ -289,7 +289,7 @@ export function languageDetector(userOptions: Partial<DetectorOptions> = {}): Mi

   validateOptions(options)

-  return async function languageDetectorMiddleware(c, next) {
+  return async function languageDetector(c, next) {
     try {
       const lang = detectLanguage(c, options)
       c.set('language', lang)
```

File: src/middleware/language/language.ts (hunk 2 above)
Comment (lord007tn) [reply src/middleware/language/language.ts:73]: I am with @sor4chi 
I believe some functions should be moved to helpers to benefit all of honor
I will work on it

**Comment by lord007tn** [comment]
i just pushed new parse-accept helpers 
and unified the logic in accept and language

my question now is for v4.7 we merge this and then turn to the docs or I do the docs to get this merged?

**Comment by sor4chi** [comment]
If you submit a PR to the honojs/website and include a link to this PR, @yusukebe should merge it at the same time as the release.

File: src/middleware/language/index.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+export type { LanguageVariables, DetectorType, CacheType }
+export {
+  languageDetector,
+  DetectorOptions,
+  detectFromCookie,
+  detectFromHeader,
+  detectFromPath,
+  detectFromQuery,
+} from './language'
+import type {} from '../..'
```
Comment (sor4chi) [inline src/middleware/language/index.ts:11]: is this required?

File: src/utils/parse-accept.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  const type = parts[0]
+  if (!type) {
+    return null
+  }
+
+  const params = parseParams(parts.slice(1))
+  const q = parseQuality(params.q)
+
+  return { type, params, q, index }
+}
```
Comment (sor4chi) [inline src/utils/parse-accept.ts:40]: I think this implementation will compile the regex object for each `parseAcceptValue` calling.
So I prefer to declare the regex in the root scope.

```suggestion
const parseAcceptValueRegex = /;(?=(?:(?:[^"]*"){2})*[^"]*$)/
const parseAcceptValue = ({ value, index }: { value: string; index: number }) => {
  const parts = value
    .trim()
    .split(parseAcceptValueRegex)
    .map((s) => s.trim())
  const type = parts[0]
  if (!type) {
    return null
  }

  const params = parseParams(parts.slice(1))
  const q = parseQuality(params.q)

  return { type, params, q, index }
}
```

File: src/utils/parse-accept.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+const parseQuality = (qVal?: string): number => {
+  if (qVal === undefined) {
+    return 1.0
+  }
+  if (qVal === '') {
+    return 1
+  }
+  if (qVal === 'NaN') {
+    return 0
+  }
```
Comment (sor4chi) [inline src/utils/parse-accept.ts:61]: q: is there any reason for using "1.0" and "0"?

File: src/middleware/language/language.ts [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+  convertDetectedLanguage?: (lang: string) => string
+  /** Enable debug logging */
+  debug?: boolean
+}
+
+export interface LanguageVariables {
+  language: string
+}
+
+const DEFAULT_OPTIONS: DetectorOptions = {
```
Comment (sor4chi) [inline src/middleware/language/language.ts:51]: It would be better to export this.

**Review by sor4chi — COMMENTED** [review]
Very good! Thank you for your support.
I have noted a few other points of concern.

**Comment by lord007tn** [comment]
> Very good! Thank you for your support. I have noted a few other points of concern.

i fixed some points you raised.

**Review by sor4chi — APPROVED** [review]
Looks good, thank you very much!

**Comment by lord007tn** [comment]
> If you submit a PR to the honojs/website and include a link to this PR, @yusukebe should merge it at the same time as the release.

I’ve added the documentation https://github.com/honojs/website/pull/569 to the honojs website. Since I’m not a technical writer by trade, please feel free to suggest improvements or refine the content for clarity and accuracy.

File: src/middleware/language/language.ts [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+    ...userOptions,
+    cookieOptions: {
+      ...DEFAULT_OPTIONS.cookieOptions,
+      ...userOptions.cookieOptions,
+    },
+  }
+
+  validateOptions(options)
+
+  return async function languageDetectorMiddleware(c, next) {
```
Comment (yusukebe) [inline src/middleware/language/language.ts:276]: Can you rename `languageDetectorMiddleware` to `languageDetector` like other middleware, e.g. [Basic Auth](https://github.com/honojs/hono/blob/main/src/middleware/basic-auth/index.ts#L86)?

File: src/utils/parse-accept.ts
Comment (yusukebe) [inline src/utils/parse-accept.ts:1]: How about renaming it to `src/utils/accept.ts`? It's simple.

File: src/utils/parse-accept.test.ts [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+      const header = 'text/html,application/json;q=0.9'
+      expect(parseAccept(header)).toEqual([
+        { type: 'text/html', params: {}, q: 1 },
+        { type: 'application/json', params: { q: '0.9' }, q: 0.9 },
+      ])
+    })
+
+    test('handles missing header', () => {
+      expect(parseAccept('')).toEqual([])
+      expect(parseAccept(undefined as any)).toEqual([])
```
Comment (yusukebe) [inline src/utils/parse-accept.test.ts:15]: Using `any` is okay, but please apply this patch to remove warnings:

```diff
diff --git a/src/utils/parse-accept.test.ts b/src/utils/parse-accept.test.ts
index a1019c92..ff97e596 100644
--- a/src/utils/parse-accept.test.ts
+++ b/src/utils/parse-accept.test.ts
@@ -12,7 +12,9 @@ describe('parseAccept Comprehensive Tests', () => {

     test('handles missing header', () => {
       expect(parseAccept('')).toEqual([])
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
       expect(parseAccept(undefined as any)).toEqual([])
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
       expect(parseAccept(null as any)).toEqual([])
     })
   })
```

File: src/middleware/language/index.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import type { LanguageVariables, DetectorType, CacheType } from './language'
+export type { LanguageVariables, DetectorType, CacheType }
+export {
+  languageDetector,
+  DetectorOptions,
+  detectFromCookie,
+  detectFromHeader,
+  detectFromPath,
+  detectFromQuery,
+  DEFAULT_OPTIONS,
```
Comment (yusukebe) [inline src/middleware/language/index.ts:10]: `DEFAULT_OPTIONS` is a type. It should not be here.

```diff
diff --git a/src/middleware/language/index.ts b/src/middleware/language/index.ts
index d225a54a..061c7538 100644
--- a/src/middleware/language/index.ts
+++ b/src/middleware/language/index.ts
@@ -1,8 +1,7 @@
-import type { LanguageVariables, DetectorType, CacheType } from './language'
-export type { LanguageVariables, DetectorType, CacheType }
+import type { LanguageVariables, DetectorOptions, DetectorType, CacheType } from './language'
+export type { LanguageVariables, DetectorOptions, DetectorType, CacheType }
 export {
   languageDetector,
-  DetectorOptions,
   detectFromCookie,
   detectFromHeader,
   detectFromPath,
```

**Comment by yusukebe** [comment]
Hi @lord007tn 

Almost all are good. I've left comments. Check them!

File: src/middleware/language/index.ts (hunk 9 above)
Comment (yusukebe) [reply src/middleware/language/index.ts:9]: @lord007tn Can you fix it?

**Review by yusukebe — CHANGES_REQUESTED** [review]

**Comment by yusukebe** [comment]
Hi @lord007tn 

Looks good to me! I'll merge this into the `next` branch for the next minor version and ship the `v4.7` soon! Thank you for the great work!
